ready for review

LuminaScript opened the pull request:
Review of the delta-kernel Tool

This pull request introduces the delta-kernel tool, which generates compile-time version update analyses between two Linux kernel versions.

LuminaScript commented:
I am creating a pull request to merge into the main repository. Thanks for your review! @mk3890 @rhodesconover8 @matthew-l-weber

matthew-l-weber commented:
@chuckwolber

mk3890 commented on README.md:
```diff
@@ -0,0 +1,105 @@
+
+## Introduction
+
+The Change Impact Analysis Tool generates a comprehensive visual report detailing changes in both header files and source code between two Linux versions (tags in the Linux kernel repository: old_tag and new_tag). This tool helps developers view updates from the old version.
```
Since we created a new name, do you want to say something like "DeltaKernel Change Impact Analysis Tool"?

LuminaScript replied:
Have renamed it as well as other similar cases.

mk3890 commented on README.md:
```diff
@@ -0,0 +1,105 @@
+# Change-impact-analysis Tool
```
Not sure what editor you are using but this file could use a markdown-link run.

If you are using VS Code there is a markdown-lint extension that will fix most things for you.

LuminaScript replied:
Okay! Just updated the readme.md after running markdown-link.

mk3890 commented on README.md:
````diff
@@ -0,0 +1,105 @@
+Execute the tool by specifying the old and new tags:
+
+```bash
+./run_tool.sh <tag1> <tag2> [-c clone_path] [-u repo_link] [-s subsystem]
````
It might be good to provide an example with the values filled out here as well.

The reason this came to mind is that is not clear to me if `repo_link` should be a URL to a repo or a path to local repo. I think from the description below that should be a URL but that is not what I would have expected by default since the tool appears to rely on building first which would mean you need a local copy.

LuminaScript replied:
Yeah! Addressed this issue with more detailed explanation and an example usage in the new commit.

mk3890 commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,163 @@
+
+set -e
+
+# check and install gcc-11 if not already installed
```
Is there a reason that gcc-11 in particular is specified? Can we not just say GCC?

The reason I call this out is that it is implied maintenance. The script will break when used against a base image that does not have access to gcc-11.

Additionally, would it make more sense to simply check for the presence of a compiler by checking if $CC is set? (I'm not sure that is a foolproof method). What if the developer is using clang/llvm to compile their kernels?

LuminaScript replied:
I believe gcc should not be hardcoded. The Linux Kernel Makefile will handle the compiler choice. I removed gcc-11 and simply ran make, and it worked.

mk3890 commented on README.md:
```diff
@@ -0,0 +1,105 @@
+
+During linux kernel compilation, `Makefile.build` calls `$K/scripts/basic/fixdep.c` to generate a .cmd file for each source that collects dependency information during compilation.
+
+This tool incorporates a modification that applies a patch (`patch.file`) to `scripts/basic/fixdep.c`, enabling it to output dependency information into a **list of header files** when building the kernel.
```
It might help to expand on why this patch is necessary?

Is something currently broken in fixdep.c? Or did we need to add something to get additional information that we needed?

LuminaScript replied:
fixdep.c generates a .cmd file for each source file. The patch collects all dependencies from each .cmd file and outputs a dependency_list.txt file, which contains the compile-time dependency file list after kernel compilation. I have added this explanation to the new README.md

mk3890 commented on build_scripts/tokenize.py:
```diff
@@ -0,0 +1,331 @@
+    args = parser.parse_args()
+
+    git_diff_report_path = args.git_diff_report
+    intermediate_file_path = args.intermediate_file
```
Is this intermediate_file something that the user would ever want or use? 

If not,

- Are we cleaning it up? (I might just have missed it)
- Suggest removing it from the argument list and just creating a random file name. There nice ways to do this in Python as I recall, Google is your friend.

LuminaScript replied:
- We are not cleaning it up. But modified the file to clean it up in the new commit. 
- Have just removed it from the arg list, let py create a random file name, and removed it right before the program exits.

mk3890 reviewed (changes requested):
Great work!

chuckwolber commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+    local patch_path="$root_path/scripts/change-impact-tools/fixdep-patch.file"
+
+    # Stash any changes if there is any
+    if ! git diff --quiet; then
+        git stash
+    fi
```
If you stash a user's changes without telling them, they are going to be surprised and perhaps a little unhappy that their repository is not in the state they expect. Instead of stashing the changes, you might consider simply exiting your script with a descriptive error if the repository is not in a state that you feel is amenable to analyzing. The user can then decide how to manage their changes and re-run your script at a later time.

LuminaScript replied:
Yes. I have fixed this problem.

chuckwolber commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+        git stash
+    fi
+
+    # Abort `git am` only if there is a patch being applied
+    if git am --show-current-patch &> /dev/null; then
+        git am --abort
+    fi
```
Same feedback as stashing. It is generally not good practice to disrupt the user like this. It may make more sense to simply exit with a descriptive error and let the user resolve the situation.

LuminaScript replied:
fixed this as well.

chuckwolber commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+        git am --abort
+    fi
+    echo "path check: $(pwd)"
+    git apply < "$patch_path"
```
Is this patch expected to apply to all possible versions of the Linux kernel? How is the user expected to handle it when this patch does not apply cleanly?

LuminaScript replied:
I am fixing this issue by instead of modifying fixdep.c, collect all .cmd generates and parse the dependency information. I will issue a new PR after this one is merged.

chuckwolber commented on README.md:
```diff
@@ -0,0 +1,115 @@
+- User customization: allows users to define the URL of the Linux root repository and specify the subsystem for analysis. (More details in [How to Use](#how-to-use))
+
+## How to use
+
```
I note that your scripts apply a patch to the kernel source. Since the source is being modified and built, you may want to be clear about the state of the user's Linux kernel source tree before and after this analysis is run. Otherwise this may surprise and confuse users who have unfinished work in their kernel source clone.

LuminaScript replied:
Same as above. Will fix this in next PR.

matthew-l-weber replied:
https://github.com/elisa-tech/delta-kernel/issues/2

chuckwolber commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+    echo "path check: $(pwd)"
+}
+
+# parse the JSON file
```
Comments should generally be very rare and, in particular, not repeat function names. Easy to read code with no comments is far better than detailed comments that will quickly go out of date as the code evolves.

LuminaScript replied:
Yes. Have already removed this comment.

chuckwolber commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+cat /etc/os-release
+
+echo "start running make"
+make 
```
I imagine you want the kernel to build to run as quickly as possible. Consider building on all available processors, like this:

```suggestion
make -j $(nproc)
```

LuminaScript replied:
Change was applied.

chuckwolber commented on build_scripts/git_shortlog.sh:
```diff
@@ -0,0 +1,20 @@
+# Fetch name email information for linux kernel contributors
+set -e
+
+TAG="$1"
```
What happens if the tag argument is not provided on the command line? You might consider checking for a value here and then calling a `usage()` function if none was provided. This is a pattern I occasionally use:

```suggestion
usage() {
    echo "Usage: $0 tag"
    exit 1
}

TAG="$1"

[[ -z ${TAG} ]] && usage
```

LuminaScript replied:
Change was applied

chuckwolber commented on fixdep-patch.file:
```diff
@@ -0,0 +1,50 @@
+diff --git a/scripts/basic/fixdep.c b/scripts/basic/fixdep.c
```
I recommend renaming this file so that the file extension is `.patch`, for example `fixdep.patch`. If this patch is only meant for a specific version (or versions) of the kernel, may also consider elaborating the name a bit further - `fixdep-v6.8-v6.9.patch`.

LuminaScript replied:
Will fix this by directly parsing the .cmd files instead of patching the source code as previously mentioned.

matthew-l-weber replied:
https://github.com/elisa-tech/delta-kernel/issues/4

chuckwolber commented on generate_build_filelists.sh:
```diff
@@ -0,0 +1,114 @@
+    clone_dir=${1:-"linux-clone"}
+
+    # Capture the absolute path of the root directory
+    root_path=$(cd ../.. && pwd)
```
See earlier suggestion on the use of `root_path` as a variable name.

LuminaScript replied:
Fixed this and all concurrent cases.

chuckwolber commented on generate_build_filelists.sh:
```diff
@@ -0,0 +1,114 @@
+
+# Ensure required arguments (tag1 and tag2) are provided
+if [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ]; then
+  echo "Usage: $0 <tag1> <tag2> <clone_path>"
```
It appears as if you already have a usage statement in the `show_help` function. Why repeat that here? Perhaps you might consider a separate `usage()` function that is called in two places to avoid the duplication?

LuminaScript replied:
I have updated the script to use show_help function instead.

chuckwolber commented on generate_web_reports.sh:
```diff
@@ -0,0 +1,34 @@
+
+# Ensure required arguments (tag1 and tag2) are provided
+if [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ]; then
+  echo "Usage: $0 <tag1> <tag2> <linux_repo_root_url>"
```
It looks like you are duplicating your usage statement here. See prior feedback on the same topic in `generate_build_filelists.sh`.

LuminaScript replied:
Thanks! I have fixed this.

chuckwolber commented on generate_web_reports.sh:
```diff
@@ -0,0 +1,34 @@
+TAG2="$2"
+REPO_URL="$3"
+
+./web_scripts/generate_data.sh "$TAG1" "$TAG2" "$REPO_URL"
```
Will this break if someone re-arranges your project repository or changes the `web_scripts` directory name?

LuminaScript replied:
- repo_url is a user-defined repository link for cloning the Linux source code directory.
- Changing the web_scripts directory name will break the functionality. However, this is not considered an issue since the user who changes the name (part of the delta-kernel source code) should also modify the path name accordingly.

chuckwolber commented on run_tool.sh:
```diff
@@ -0,0 +1,92 @@
+REPO_URL="https://github.com/gregkh/linux"
+SUBSYS=""
+
+print_usage() {
+  echo "Usage: $0 <tag1> <tag2> [-c clone_path] [-u repo_link] [-s subsystem]"
+}
```
It looks like this usage statement is duplicated earlier in this script. Would it be better to have the earlier function call this function? If so, you would want this function to appear earlier in the file so that it is available when it is needed.

LuminaScript replied:
Thanks for pointing this out. I have moved this function earlier and used it in later code.

chuckwolber commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+}
+
+
+TAG1="$1"
+TAG2="$2"
+SUBSYS="$3"
```
What happens when these arguments are missing?

LuminaScript replied:
Those arguments are first passed into the run_tool script, which outputs a usage message if any arguments are missing. Then, run_tool passes these arguments into generate_build_filelists, which subsequently passes them into build_collect_diff. Therefore, I removed the redundant argument checks from the other scripts.

chuckwolber commented on README.md:
```diff
@@ -0,0 +1,115 @@
+
+- [Introduction](#introduction)
+- [Innovation](#innovation)
+- [How to Use](#how-to-use)
```
I recommend adding an `Assumptions` section that describes all of the assumptions you are making about the state of the user's Linux kernel source repository clone, how that clone will be used by your script, and what state it will be in when your script is complete. You may also want to provide some manual cleanup guidance in case something goes wrong and the user needs to "reset" themself to a known good configuration.

LuminaScript replied:
I added sections inside the "How to Use" section explaining both how the tool runs and the stages of cloning the Linux repository after execution. I also added manual cleanup guidance and instructions on resetting the Linux repository in case of git conflicts.

matthew-l-weber replied:
https://github.com/elisa-tech/delta-kernel/issues/2

chuckwolber commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+}
+
+# clean up the working directory
+cleanup_working_directory() {
+    git reset --hard
+    git clean -fdx
+}
```
This is very concerning to me. What happens if the user has important work in progress when they run this script? If you do not expect them to have work in progress, should you note that assumption in your documentation? And what are the chances the user will miss the documentation and violate your assumption?

LuminaScript replied:
This is fine because the cleanup_working_directory is run in the cloned linux repository after we want to checkout a new tag (to make sure no conflict occurs). I will leave comment/documentation to illustrate this. Thanks! @chuckwolber

chuckwolber replied:
Yes, that is a good point. It looks like you are using the cloned kernel as your "work area", so there should not be any problems doing a hard reset at the end of your analysis. I noted your comment/documentation addressing this as well.

However, upon thinking about it further, since you are using the Linux clone as a "work area", I wonder if you could simplify things in the following way?

In `generate_build_filelists` you have the following code:

```
    if [ ! -d "$clone_kernel_repo_path/.git" ]; then
        git clone "$kernel_repo_path" "$clone_kernel_repo_path"
    fi
```

I think if you were to modify it slightly to do an automatic cleanup if the clone already exists, you could avoid the need to document cleanup instructions. Even if your script failed to run and left the clone in an inconsistent state, you would always ensure it is in a perfectly consistent state when you start up the next time.

```
    if [ ! -d "$clone_kernel_repo_path/.git" ]; then
        git clone "$kernel_repo_path" "$clone_kernel_repo_path"
    else
        git -C "$clone_kernel_repo_path" reset --hard
        git -C "$clone_kernel_repo_path" clean -fdx
    fi
```

You could probably get rid of `cleanup_working_directory()` in `build_collect_diff`, as well as the code in `apply_patch()` that checks for unstashed changes and `--show-current-patch`.

matthew-l-weber replied:
https://github.com/elisa-tech/delta-kernel/issues/2

matthew-l-weber commented:
@chuckwolber @mk3890   I'll take on most of these changes when I'm back.  I wanted to have @LuminaScript publish a draft before she's done this next week!

LuminaScript commented:
Thanks for the comments @mk3890 ! I will fix this by Monday. @matthew-l-weber Matt Kelly is reviewing my pull request and we will merge it once all comments are resolved.

LuminaScript commented on build_scripts/tokenize.py:
```diff
@@ -0,0 +1,334 @@
+"""
```
Okay! Fixed.

LuminaScript commented on build_scripts/parse_json.py:
```diff
@@ -0,0 +1,75 @@
+"""
```
Okay! Fixed.

LuminaScript commented on build_scripts/parse_dep_list.py:
```diff
@@ -0,0 +1,71 @@
+"""
```
Okay! Fixed.

LuminaScript commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+#!/bin/bash
```
Okay! Fixed.

LuminaScript commented on build_scripts/git_shortlog.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Okay! Fixed.

LuminaScript commented on generate_build_filelists.sh:
```diff
@@ -0,0 +1,114 @@
+#!/bin/bash
```
Okay! Fixed.

LuminaScript commented on generate_web_reports.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Okay! Fixed.

LuminaScript commented on run_tool.sh:
```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
```
Okay! Fixed.

LuminaScript commented on web_scripts/generate_data.sh:
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
```
Okay! Fixed.

LuminaScript commented on README.md:
```diff
@@ -0,0 +1,115 @@
+
+The idea of generating a web display for Linux kernel version change impact analysis is inspired by [Cregit](https://github.com/cregit/cregit). This tool innovates on Cregit by:
+
+- Considering the extensive code space the Linux kernel deals with, it provides a compile-time analysis instead of a static analysis of the commit history of the Linux kernel, presenting changes only in files used during compilation. 
```
The tool does not allow user to define kernel configuration. "compile-time" means it do a version analysis (like git diff) only for files that are used to compile the kernel (instead of the whole Linux source code space). I have modified that documentation for more clarification.

LuminaScript commented on README.md:
```diff
@@ -0,0 +1,115 @@
+
+## How to use
+
+To utilize this tool in your Linux environment (compatible with Ubuntu and Debian), follow these steps:
```
I have specified the version to Ubuntu 22.04 (the only one tested with) and have updated it in the README.md.

LuminaScript commented on README.md:
```diff
@@ -0,0 +1,115 @@
+- Enabling comparison between two specific tags/releases in the Linux kernel, highlighting all newly added and deleted lines. This provides a clear layout of differences between the tags. While Cregit organizes information by files and embeds the latest commit details in each line/token, it does not support direct comparison between two tags.
+- User customization: allows users to define the URL of the Linux root repository and specify the subsystem for analysis. (More details in [How to Use](#how-to-use))
+
+## How to use
```
No. the linux-clone is cloned from the github default linux repo.

chuckwolber replied:
That should be fine. What made you choose `olddefconfig` instead of `defconfig`?

The `olddefconfig` make target is meant to be used when you want to update an existing `.config`, but it does not look like you are updating any existing `.config` file. You can still use `olddefconfig` even without a `.config` because `olddefconfig` is equivalent to `defconfig` when no `.config` is found. But `olddefconfig` behaves differently compared to `defconfig` when a `.config` file is found.

This behavior difference may cause problems in the future if you do not intend to update pre-existing `.config` files.

LuminaScript replied:
Yes I think using `make defconfig` is safer. No particular reason to use `oldefconfig` out of all `make *config`. I have updated the code with the changes.

LuminaScript commented on README.md:
```diff
@@ -0,0 +1,115 @@
+
+## Structure of the Tool
+
+The tool operates through a structured process to generate a comprehensive change impact analysis report. Here's a detailed breakdown of its operation:
```
I think the title should be "operation stages". It goes through how the tool operates in detailed. Have applied this changes.

matthew-l-weber replied:
updated in PR2 - https://github.com/elisa-tech/delta-kernel/pull/3/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R117

LuminaScript commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+
+echo "start running make"
+make 
+echo "finished compile kernel using gcc 11"
```
No. It was a mistake. I have removed it. Thanks for pointing it out!

LuminaScript commented on build_scripts/build_collect_diff.sh:
```diff
@@ -0,0 +1,157 @@
+# safely apply a patch to linux kernel
+apply_patch() {
+    # shellcheck disable=SC2154
+    local patch_path="$root_path/scripts/change-impact-tools/fixdep-patch.file"
```
It is the last case as it refers to the directory the linux kernel source code is kept. I have changed the variable name to kernel_repo_path consistently in all scripts.

LuminaScript commented:
@chuckwolber The latest commit should have addressed all the concerns you pointed out in the comments. Thank you for reviewing my code! Please let me know if any further changes are needed.

chuckwolber commented on run_tool.sh:
```diff
@@ -0,0 +1,92 @@
+TAG2="$2"
+
+CLONE_PATH="linux-clone"
+REPO_URL="https://github.com/gregkh/linux"
```
Did you mean to say that you were using the GitHub clone of the stable kernel? The canonical stable kernel is found here `https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git`. The one you are using (`https://github.com/gregkh/linux`) is just a mirror of the stable kernel.

It is perfectly reasonable to do what you are doing, I just wanted to make sure you understood the distinction.

LuminaScript replied:
Oh yes the default is `https://github.com/gregkh/linux` but the user can use `-u` flag to pass in the canonical stable kernel url as well.

chuckwolber commented on README.md:
````diff
@@ -0,0 +1,151 @@
+**Example Usage**:
+
+```bash
+./run_tool "v5.15" "v5.15.100" -c "linux-clone" -u "https://github.com/torvalds/linux" -s "security"
````
It looks like your tags got out of sync with your comment. 

```suggestion
./run_tool "v6.8" "v6.9" -c "linux-clone" -u "https://github.com/torvalds/linux" -s "security"
```

LuminaScript replied:
Actually I have updated all to v5.15 vs v5.15.100, for later version might run into conflict issue when applying the patch.

chuckwolber commented on build_scripts/build_collect_diff:
```diff
@@ -0,0 +1,158 @@
+    local input_path="$kernel_repo_path/scripts/change-impact-tools/build_data/compile_commands.json"
+    local output_path="$kernel_repo_path/scripts/change-impact-tools/build_data/sourcefile.txt"
+
+    ./"$python_path" "$cloned_repo_name" "$input_path" "$output_path"
```
I was not able to run your script until I made this change.

```suggestion
    "$python_path" "$cloned_repo_name" "$input_path" "$output_path"
```

chuckwolber commented on build_scripts/build_collect_diff:
```diff
@@ -0,0 +1,158 @@
+    # Generate compiled source files list
+    local json_output_path="$kernel_repo_path/scripts/change-impact-tools/build_data/compile_commands.json"
+    echo "path check: $(pwd)"
+    ./scripts/clang-tools/gen_compile_commands -o "$json_output_path"
```
The script in the Linux kernel source tree still has the `.py` extension. I had to make this change to get your script to work properly.

```suggestion
    ./scripts/clang-tools/gen_compile_commands.py -o "$json_output_path"
```

chuckwolber commented on build_scripts/build_collect_diff:
```diff
@@ -0,0 +1,158 @@
+    local dep_path="dependency_file.txt"
+    local python_tool_path="$kernel_repo_path/scripts/change-impact-tools/build_scripts/parse_dep_list"
+
+    ./"$python_tool_path" "$dep_path" "$output_json" "$output_list"
```
I was not able to run your script until I made this change.

```suggestion
    "$python_tool_path" "$dep_path" "$output_json" "$output_list"
```

chuckwolber commented on build_scripts/build_collect_diff:
```diff
@@ -0,0 +1,158 @@
+
+set -e
+
+# check and install gcc-11 if not already installed
+install_package_safe() {
+    if ! command -v libssl-dev &> /dev/null; then
+        sudo apt-get update
+        sudo apt-get install -y libssl-dev
+    else
+        echo "libssl-dev is already installed."
+    fi
+}
```
This function does not actually do anything useful. I would recommend removing it.

In particular, you probably do not need `libssl-dev` to compile the kernel, and `libssl-dev` is not a valid command, so this will always run the `apt` commands, even when `libssl-dev` is properly installed.

LuminaScript replied:
Okay  I have removed the function!

matthew-l-weber replied:
I think it was ever since ~4.x libssl-dev is needed as a dependency because of building signing tools.  I'll make a ticket to fix this host dependency as part of containerizing the script execution so dependencies are stable.

chuckwolber commented on build_scripts/build_collect_diff:
```diff
@@ -0,0 +1,158 @@
+git checkout "$TAG1"
+# Preparation before running make
+apply_patch
+install_package_safe
```
Remove this line once you remove the `install_package_safe()` function.

LuminaScript replied:
Have just removed it as well!

chuckwolber commented on generate_build_filelists:
```diff
@@ -0,0 +1,113 @@
+# Retrieve and tokenize commit info per added line
+echo "tokenizing each commit line ..."
+git checkout "$TAG2"
+./"$curr_dir"/build_scripts/tokenize "$curr_dir/build_data/filtered_diff_header.txt" "$curr_dir/build_data/tokenize_header.json" "$TAG1" "$TAG2"
+./"$curr_dir"/build_scripts/tokenize "$curr_dir/build_data/filtered_diff_source.txt" "$curr_dir/build_data/tokenize_source.json" "$TAG1" "$TAG2"
```
I was not able to run your script until I made this change.

```suggestion
"$curr_dir"/build_scripts/tokenize "$curr_dir/build_data/filtered_diff_header.txt" "$curr_dir/build_data/tokenize_header.json" "$TAG1" "$TAG2"
"$curr_dir"/build_scripts/tokenize "$curr_dir/build_data/filtered_diff_source.txt" "$curr_dir/build_data/tokenize_source.json" "$TAG1" "$TAG2"
```

chuckwolber reviewed (changes requested):
While testing your tool I found a few things that you may need to fix.

matthew-l-weber commented:
I found a dependency bug; tooling needs to run inside a container.  (will make a ticket)

```
Parsing git line info in file: security/selinux/include/security.h
Parsing git line info in file: security/selinux/include/policycap_names.h
Parsing git line info in file: security/selinux/ss/sidtab.h
Diff info stored in /tmp/tmpy4q7b9ra
starting tokenize function test
Result exported to tokenize.json
Traceback (most recent call last):
  File "/wip/delta-kernel/linux/scripts/change-impact-tools/build_scripts/tokenize", line 335, in <module>
    os.remove(intermediate_file_path)
NameError: name 'os' is not defined

```

matthew-l-weber commented on build_scripts/tokenize:
```diff
@@ -0,0 +1,335 @@
+import json
+import argparse
+from datetime import datetime
+import tempfile
```
Needs `import os` as well for the use of `os.remove`

matthew-l-weber replied:
Fixed in PR2 - https://github.com/elisa-tech/delta-kernel/pull/3/files#diff-2a307bf1d08e3f4518d4887c595a9ff4eabd51094dab87f465790e1162fa3c17R18

matthew-l-weber commented:
Moved a rebased set of changes with some fixups to PR2 - https://github.com/elisa-tech/delta-kernel/pull/3/files

matthew-l-weber commented:
Thank you @LuminaScript  for the great summer!
